src/Agent.Worker/Build/BuildServer.cs

```diff
@@ -12,22 +11,62 @@
 
 namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
 {
-    public class BuildServer
+    [ServiceLocator(Default = typeof(BuildServer))]
+    public interface IBuildServer : IAgentService
     {
-        private readonly Build2.BuildHttpClient _buildHttpClient;
-        private Guid _projectId;
+        Task ConnectAsync(VssConnection jobConnection);
+        Task<Build2.BuildArtifact> AssociateArtifactAsync(
+            int buildId,
+            Guid projectId,
+            string name,
+            string jobId,
+            string type,
+            string data,
+            Dictionary<string, string> propertiesDictionary,
+            CancellationToken cancellationToken = default(CancellationToken));
+        Task<Build2.Build> UpdateBuildNumber(
+            int buildId,
+            Guid projectId,
+            string buildNumber,
+            CancellationToken cancellationToken = default(CancellationToken));
+        Task<IEnumerable<string>> AddBuildTag(
+            int buildId,
+            Guid projectId,
+            string buildTag,
+            CancellationToken cancellationToken = default(CancellationToken));
+    }
+
+    public class BuildServer : AgentService, IBuildServer
+    {
+        private VssConnection _connection;
+        private Build2.BuildHttpClient _buildHttpClient;
 
-        public BuildServer(VssConnection connection, Guid projectId)
+        public async Task ConnectAsync(VssConnection jobConnection)
         {
-            ArgUtil.NotNull(connection, nameof(connection));
-            ArgUtil.NotEmpty(projectId, nameof(projectId));
+            _connection = jobConnection;
+            int attemptCount = 5;
+            while (!_connection.HasAuthenticated && attemptCount-- > 0)
+            {
+                try
+                {
+                    await _connection.ConnectAsync();
+                    break;
+                }
+                catch (Exception ex) when (attemptCount > 0)
```

Comment:
Do we need the `when`?

Reply:
This code was copied from the other `IAgentService` classes that use a `VssConnection`. From what it looks like, the `when` isn't necessary here, but I'd rather not change that here in this PR without changing other places with this pattern.